olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,191 @@
+
+    def get_model_config_by_hash_key(self, output_model_hash: str):
+        model_config_blob = f"{output_model_hash}/model_config.json"
+        model_config_path = Path(self.cache_dir) / output_model_hash
+        model_config_path.mkdir(parents=True, exist_ok=True)
+        model_config_path = model_config_path / "model_config.json"
+
+        blob_client = self.container_client.get_blob_client(model_config_blob)
+        logger.info("Downloading %s to %s", model_config_blob, model_config_path)
+
+        with open(model_config_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
+            download_file.write(download_stream.readall())
+
+        # model config
+        with open(model_config_path) as file:
+            model_config_dict = json.load(file)
+            return ModelConfig.from_json(model_config_dict)
+
+    def upload_model_to_cloud_cache(self, output_model_hash: str, output_model_config: ModelConfig):
+        logger.info("Uploading model to cloud cache.")
+        model_path = output_model_config.config.get("model_path")
+
+        if model_path is None:
+            logger.error("Model path is not found in the output model config. Upload failed.")
+            return
+
+        model_path = Path(model_path)
+        model_config_copy = deepcopy(output_model_config)
+        model_config_copy.config["model_path"] = None
+
+        # upload model config file
+        model_config_bytes = json.dumps(model_config_copy.to_json()).encode()
+        with io.BytesIO(model_config_bytes) as data:
+            self.container_client.upload_blob(f"{output_model_hash}/model_config.json", data=data, overwrite=False)
+
+        # upload model file
+        model_blob = str(Path(output_model_hash) / "model")
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            model_path_file = Path(temp_dir) / "model_path.json"
+            with open(model_path_file, "w") as file:
+                file.write({"model_path": model_path.name})
+            with open(model_path_file, "rb") as data:
+                self.container_client.upload_blob(f"{model_blob}/model_path.json", data=data, overwrite=False)
+
+        if not model_path.is_dir():
+            with open(model_path, "rb") as data:
+                self.container_client.upload_blob(f"{model_blob}/{model_path.name}", data=data, overwrite=False)
+        else:
+            self._upload_dir_to_blob(model_path, f"{model_blob}")
+
+    def _upload_dir_to_blob(self, dir_path, blob_folder_name):
+        for item in dir_path.iterdir():
+            if item.is_dir():
+                self._upload_dir_to_blob(item, f"{blob_folder_name}/{item.name}")
+            else:
+                blob_name = f"{blob_folder_name}/{item.name}"
+                with open(item, "rb") as data:
+                    self.container_client.upload_blob(name=blob_name, data=data, overwrite=False)
```

Comment:
nit: Good opportunity to parallelize this with multiprocessing. Some of these files can be very large and uploading them to remote can be time consuming. Uploading them in parallel will be a low hanging fruit to optimize it.

Reply:
Sounds good to me. I can improve this later in a different PR.

Reply:
https://aiinfra.visualstudio.com/Model%20optimization%20Toolkit/_workitems/edit/39356